### ISSUE TITLE
Refactoring the task queue and scheduler

### ISSUE THREAD

**Issue by Kerollmops** [meilisearch/meilisearch#2725]
In this issue, we simply list the requirements for a new task queue. The current one is pretty complex and adding new functionalities is cumbersome. We would like to design a new scheduler system that fixes and improves over the current one.

For that, we must evaluate the time we need to take to develop most of the features below:

- [x] Durability of the tasks once an answer has been sent to the user.
- [x] Disable the auto-batching system.
- [ ] Cancel one or _multiple_ enqueued or processing tasks.
- [ ] Remove already processed tasks from the task store.
   - [ ] don't delete tasks with an id higher or equal to the deleting task
   - [ ] use bitmaps for the matched task ids
- [ ] Swapping indexes.
- [ ] Support different task types:
    - [ ] Import dumps at startup time.
    - [ ] Export dumps i.e. export the tasks queue.
    - [x] Document addition (replace or update)
    - [x] Document deletion.
    - [x] Clear all documents.
    - [x] Update the settings of an index.
    - [x] Index update e.g. Create an Index, change an index name, delete an index.
    - [ ] Cancel a processing dump.
- [ ] Auto-batching of enqueued tasks:
    - [x] Document addition.
    - [x] Document updates.
    - [ ] Settings.
    - [ ] Document deletion.
- [ ] Prioritize and reorder tasks i.e. Index deletion, Delete all the documents.
- [x] Dump the tasks queue itself.
- [x] Always accept new tasks without blocking.
- [x] Fairly share the loads over the different indexes e.g. Always process the index queue with the lowest id.
- [ ] Easily testable.
- [ ] Well tested i.e. tasks reordering, tasks prioritizing, use atomic barriers to block the tasks for tests.
- [ ] Realtime progressing status e.g. Websocket events.
- [x] List and filter tasks:
    - [x] Paginate the tasks.
    - [x] Filter by index name.
    - [x] Filter on the status, the enqueued, processing, and finished tasks.
    - [x] Filter on the type of task.

**Comment by Kerollmops** [comment]
We have done a meeting yesterday (@irevoire, @loiclec, and myself) to evaluate the time it would take to design the new scheduler designed to support all of the features listed above and with the less code complexity possible.

We decided that we will use LMDB as the task and index name mapper store. The task ids are incremental `u32`s. As a value, we have all the information about the task (see below). When the task type is a document addition, a file with the content is created and stored outside of the LMDB store.

We must be using a [`spawn_blocking`](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html) to correctly inform the HTTP async reactor that we are in fact waiting and blocking.

The order of the tasks is defined by the acquiring of the write lock into LMDB, it isn't determined at the moment we receive the first byte from the network. It is because we will write the update's content in a file **before** acquiring the write transaction to LMDB and defining the task id.

```rust
/// This struct must be ran only once, it must be a singleton.
struct TaskScheduler {
    should_abort: Arc<AtomicBool>,
    processing_tasks: Mutex<Vec<TaskId>>,
}

impl TaskScheduler {
    pub async fn register(&self, task: TaskWithContent) -> Result<TaskView> {
        let mut wtxn = todo!();

        task.content_file.persist(content_file_uuid)?;
        let task_id = self.disk.get_next_task_id()?;
        let task = Task { content_file_uuid, .. };
        self.disk.store(task)?;

        wtxn.commit()?;
        // please recover the content file if an error occurred

        self.notify_task_registered();
    }

    fn get_next_batch(&self) -> Result<BatchOperation> {
        // we look at the first enqueued task id,
        // we get the list of all the tasks of an index.

        let dumps = self.prioritized_tasks();
        let enqueued_tasks = self.all_tasks_of_index(first_enqueued_id).collect();
        let operation: BatchOperation = auto_batcher_priorizer(enqueued_tasks);
        return operation;
    }

    fn abort_processing_tasks(&self, tasks: impl Iterator<Item = TaskId>) {
        // we abort the task that is currently running **only**
        // if one of these is contained in this list.
    }

    /// This worker function must be run in a different thread and must be run only once.
    fn scheduler_worker(&self) {
        loop {
            let batch_operation: BatchOperation = self.get_next_batch().unwrap();

            // do the work but... where is the index ???

            match batch_operation {
                BatchOperation::IndexSwap { a, b } => {
                    self.store.index_swap(a, b)?;
                }
            }

            for (task_id, result) in task_results {
                self.store.write_as_finished(task_id, result)?;
            }

            self.notify_finished();
            self.wait_to_be_notified();
        }
    }
}
```

## Aborting multiple tasks

Aborting one or multiple tasks is also a task that is enqueued and prioritized. We obviously only stop the current indexation process if a task that should be aborted is currently running.
 - The abortion task is written in the store with a new task id.
 - We instantly inform milli to stop the current indexation by using [an `AtomicBool`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicBool.html).
 - Making milli to return an `IndexationAbortion` error, the scheduler can react to it by ignoring it
 - and polling the tasks to process, finding the abortion task as it is prioritized and processed first
 - It processes the abortion task by marking them as canceled and removing them from the enqueued database.

## Prioritizing and auto-batching tasks

We decided that the system that was prioritizing and deciding how to auto-batch tasks could return a `BatchOperation` result. We would like, for example, to be able to skip all _document additions_ that are before any _clear all document_ operation. For that, the system must be aware of all the enqueued task types.

The prioritizer and auto-batching system receives the task id associated with the task type and will be able to return a `BatchOperation` result, the tasks to process can be scattered which is very useful to prioritize tasks and also mark useless tasks as done in advance.

```rust
enum BatchOperation {
    ClearAllDocuments {
        /// Can contain all of the _document additions_ that were
        /// enqueued before a _clear all documents_ task.
        tasks: Vec<TaskId>,
    },
    DocumentAddition {
        tasks: Vec<TaskId>,
        content_files: Vec<File>,
        method: enum { Update, Replace },
    }
    DocumentDeletion {
        tasks: Vec<TaskId>,
        documents_ids: Vec<DocumentId>,
    }
    IndexSwap {
        task: TaskId,
        index_name_a: String,
        index_name_b: String,
    },
}

impl BatchOperation {
    /// Returns the list of tasks that will be marked as finished
    /// once the indexing process is done.
    fn tasks_to_process(&self) -> Vec<TaskId>;
}
```

## About the index task store and task types

We decided that we would put all of the tasks into a single database and put references to them from different databases, this will ease the filtering and pagination system a lot. When a task is newly created we put it into the _all tasks_ database and also push the task id into the _enqueued tasks_ database. When a task is marked as finished we move the task id from the enqueued to the _finished tasks_ database and update the `InternalTask` database accordingly.

We need to think more about the index name mapper, it must be designed to accept basic operations on the indexes: renaming, swapping, deletions, and creations.

```rust
struct IndexTaskStore {
    all_tasks: Database<TaskId, InternalTask>,
    enqueued_tasks: Database<TaskId, ()>,
    finished_tasks: Database<TaskId, ()>,

    index_name_mapper: Database<IndexName, IndexUuid>,
    index_tasks: Database<IndexUuid, Set<TaskId>>,
}
```

The `TaskWithContent` is given by the user of the scheduler, and is internally processed by the scheduler: the content files are assigned a UUID and moved to the right place. It is converted into an `InternalTask` with only the information required. When a task must be returned to the user we return a `TaskView` which only contains what we want to show.

```rust
enum TaskWithContent {
    DumpImport { input: PathBuf },
    DumpExport { output: PathBuf },
    DocumentAddition { index_name: String, content_file: File },
    DocumentDeletion { index_name: String, documents_ids: Vec<String> },
    ClearAllDocuments { index_name: String },
    Settings { index_name: String, new_settings: Settings },
    RenameIndex { index_name: String, new_name: String },
    CreateIndex { index_name: String, primary_key: Option<String> },
    DeleteIndex { index_name: String },
    SwapIndex { lhs: String, rhs: String },
    CancelTask { tasks: Vec<TaskId> },
}

enum InternalTask {
    DumpImport { input: PathBuf },
    DumpExport { output: PathBuf },
    DocumentAddition { index_name: String, content_file_uuid: Uuid },
    DocumentDeletion { index_name: String, documents_ids: Vec<String> },
    Settings { index_name: String, new_settings: Settings },
    RenameIndex { index_name: String, new_name: String },
    CreateIndex { index_name: String, primary_key: Option<String> },
    DeleteIndex { index_name: String },
    SwapIndex { index_name_a: String, index_name_b: String },
    CancelTask { tasks: Vec<TaskId> },

    // each variant has those fields too
    enqueued_at: Datetime,
    processed_at: Option<DateTime>,
    finished_at: Option<DateTime>,
    error: Option<String>,
    status: Option<JsonValue>,
}

struct TaskView {
    task_id: TaskId,
    status: TaskStatus,
    enqueued_at: DateTime,
    processed_at: Option<DateTime>,
}
```

**Comment by Kerollmops** [comment]
We have done a follow-up meeting today with @irevoire, @loiclec, and @gmourier on the index scheduler (the new name ✨).

### How do we give access to the indexes

We need to provide the indexes of milli to the HTTP handler, and for that, we must keep them open in the scheduler. We thought that a simple `RwLock<HashMap<IndexUuid, Index>>` would do the job.

The swap operation would not even touch the hashmap itself we just need to swap the name index-UUID relation on disk to persist the operation. Every time an index is requested we just resolve the index-UUID on disk by using the index name. That's all.

```rust
struct IndexScheduler {
    index_map: RwLock<HashMap<IndexUuid, Index>>,

    all_tasks: Database<TaskId, InternalTask>,
    priority_enqueued_tasks: RoaringBitmap<TaskId>,
    enqueued_tasks: RoaringBitmap<TaskId>,
    succeeded_tasks: RoaringBitmap<TaskId>,
    failed_tasks: RoaringBitmap<TaskId>,

    index_name_mapper: Database<IndexName, IndexUuid>,
    index_tasks: Database<IndexName, RoaringBitmap<TaskId>>,
}

impl IndexScheduler {
    pub fn index(&self, name: &str) -> Result<Index> {
        let rtxn = todo!();
        let uuid = self.index_name_mapper.get(rtxn, name)?;
        // if not found open the it
        self.index_map.read().get(uuid)?.unwrap().clone()
    }

    fn swap_indexes(&self, namea: &str, nameb: &str) -> Result<()> {
        let mut wtxn = todo!();
        let uuida = self.index_name_mapper.get(&mut wtxn, namea)?;
        let uuidb = self.index_name_mapper.get(&mut wtxn, nameb)?;

        // swap
        self.index_name_mapper.put(&mut wtxn, namea, uuidb)?;
        self.index_name_mapper.put(&mut wtxn, nameb, uuida)?;

        wtxn.commit()
    }
}
```

### About processing and prioritizing tasks

We had some concerns about how we could be sure to prioritize tasks correctly. We decided that we needed two kinds of prioritizers, a global one and an index-level one:
 - The global one will retrieve the important tasks to do first, i.e., Cancel tasks, and dumps.
 - The index-level one will just manage the tasks of a single index and will auto-batch them.

We also decided that task cancelation must be processed one by one and must not be auto-batched. They will be processed in reverse order: the latest enqueued will be the first processed. This assures a correct and logical way of processing cancelation.

Swap indexes operations must be pushed to both indexes at the same time. This way, they are visible in both tasks queue for both indexes. We also want the history of operations to be kept for indexes. Therefore, when a swap occurs, processed tasks must be swapped with the other index.

```rust
enum PrioritizedTaskType {
    DumpImport,
    DumpExport,
    SnapshotExport,
    CancelTask,
}

enum IndexTaskType {
    DocumentAdditionUpdate,
    DocumentAdditionReplace,
    DocumentDeletion,
    ClearAllDocuments,
    Settings,
    RenameIndex,
    CreateIndex,
    DeleteIndex,
    SwapIndex,
}

/// Only processes global operations and returns a batch operation that the scheduler must process.
fn global_prioritizer(tasks: &[(TaskId, PrioritizedTaskType)]) -> Option<BatchOperation>;

/// Only processes index-level operations. Try to auto-batch tasks to fasten the processing.
fn index_prioritizer_auto_batcher(tasks: &[(TaskId, IndexTaskType)]) -> Option<BatchOperation>;
```

### (I forgot what I was thinking about...)

### Sharing the knowledge

To help other teams and people understand what we are capable of we should:
 - Create a schema of which task types are cancellable when in progress.
 - Create a list of actions and history of the index tasks lists when an index swap is done.
